File: frontend/src/api/scrape-coupang.py
import requests
import re
from bs4 import BeautifulSoup

url_start = 'https://www.coupang.com/np/search?component=&q='
url_end = '&channel=user'

headers = {
  "authority": "www.coupang.com",
  "method": "GET",
  "path": "/np/search?component=&q=socks&channel=user",
  "scheme": "https",
  "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7",
  "accept-encoding": "gzip, deflate, br, zstd",
  "accept-language": "en-US,en;q=0.9",
  "cache-control": "max-age=0",
  "priority": "u=0, i",
  "sec-ch-ua": "\"Not)A;Brand\";v=\"99\", \"Google Chrome\";v=\"127\", \"Chromium\";v=\"127\"",
  "sec-ch-ua-mobile": "?0",
  "sec-ch-ua-platform": "macOS",
  "sec-fetch-dest": "document",
  "sec-fetch-mode": "navigate",
  "sec-fetch-site": "none",
  "sec-fetch-user": "?1",
  "upgrade-insecure-requests": "1",
  "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36"
}

class CoupangScraper:
  def get_page_content(self, url):
    response = requests.get(url, headers=headers, timeout=10)
    if response.status_code == 200:
      return response.text
    return None
  def parse_products(self, page_content):
    image_urls = []
    soup = BeautifulSoup(page_content, 'html.parser')
    product_ul = soup.find('ul', id='productList')
    product_lis = product_ul.find_all('li')
    top_product_lis = []
    for product_li in product_lis:
      product_li_a = product_li.find('a', recursive=False)
      if not product_li_a:
        break
      product_li_span = product_li_a.find('span', recursive=False)
      if product_li_span and 'number' in product_li_span.get('class', []):
        top_product_lis.append(product_li)
    print("started printing: ")
    for product_li in top_product_lis:
      a = product_li.find('a', recursive=False)

      # gets the href of the product link
      href = a.get('href')
      product_url = f'https://www.coupang.com{href}'
      
      dl = a.find('dl', recursie=False)
      img = dl.find('img')
      src = img.get('src')
      src = re.sub(r'^[^a-zA-Z0-9]+', '', src)

      dd = dl.find('dd', recursive=False)
      div = dd.find('div', recursive=False)
      div_name = div.find('div', class_="name", recursive=False)
      name = div_name.get_text()
      
      div_price_area = div.find('div', class_="price-area", recursive=False)
      div_price_wrap = div_price_area.find('div', class_="price-wrap", recursive=False)
      div_price = div_price_wrap.find('div', class_="price", recursive=False)
      em_sale = div_price.find('em', class_="sale", recursive=False)
      strong_price_value = em_sale.find('strong', class_="price-value", recursive=False)
      price = strong_price_value.get_text()

      image_urls.append({
        "product_url": product_url,
        "src": src,
        "name": name,
        "price": price
      })
    return image_urls
  def make_request(self, item):
    url = url_start + item + url_end
    page_content = self.get_page_content(url)
    if page_content:
      return self.parse_products(page_content)
    return None


def main():
  coupang_scraper = CoupangScraper()

if __name__ == "__main__":
  main()

File: frontend/src/api/scrape-coupang.ts
// scrape-coupang.ts

import { useQuery } from '@tanstack/react-query';
import { categories } from './fetch-categories';

const URL_START = 'https://www.coupang.com/np/search?component=&q=';
const URL_END = '&channel=user';

const headers = {
  authority: 'www.coupang.com',
  method: 'GET',
  path: '/np/search?component=&q=socks&channel=user',
  scheme: 'https',
  accept:
    'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7',
  'accept-encoding': 'gzip, deflate, br, zstd',
  'accept-language': 'en-US,en;q=0.9',
  'cache-control': 'max-age=0',
  priority: 'u=0, i',
  'sec-ch-ua':
    '"Not)A;Brand";v="99", "Google Chrome";v="127", "Chromium";v="127"',
  'sec-ch-ua-mobile': '?0',
  'sec-ch-ua-platform': 'macOS',
  'sec-fetch-dest': 'document',
  'sec-fetch-mode': 'navigate',
  'sec-fetch-site': 'none',
  'sec-fetch-user': '?1',
  'upgrade-insecure-requests': '1',
  'user-agent':
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36',
};

interface ProductInfo {
  product_url: string;
  src: string;
  name: string;
  price: string;
}

const getPageContent = async (url: string): Promise<string> => {
  const response = await fetch(url, { headers });
  if (!response.ok) {
    throw new Error(`HTTP error! status: ${response.status}`);
  }
  return response.text();
};

const parseProducts = (pageContent: string): ProductInfo[] => {
  const parser = new DOMParser();
  const doc = parser.parseFromString(pageContent, 'text/html');
  const productUl = doc.getElementById('productList');
  if (!productUl) return [];

  const topProductLis = Array.from(productUl.querySelectorAll('li')).filter(
    (li) => li.querySelector('a > span.number')
  );

  return topProductLis
    .map((productLi) => {
      const a = productLi.querySelector('a');
      if (!a) return null;

      const href = a.getAttribute('href');
      const productUrl = `https://www.coupang.com${href}`;

      const img = a.querySelector('dl img');
      const src = img?.getAttribute('src')?.replace(/^[^a-zA-Z0-9]+/, '') || '';

      const name = a.querySelector('dl dd div.name')?.textContent || '';
      const price =
        a.querySelector(
          'dl dd div.price-area div.price-wrap div.price em.sale strong.price-value'
        )?.textContent || '';

      return {
        product_url: productUrl,
        src,
        name,
        price,
      };
    })
    .filter((product): product is ProductInfo => product !== null);
};

const fetchCoupangProducts = async (q: string): Promise<ProductInfo[]> => {
  const url = `${URL_START}${q}${URL_END}`;
  const pageContent = await getPageContent(url);
  return parseProducts(pageContent);
};

export const useCoupangScraper = () => {
  const q = categories.items[1].name.toLowerCase().replace(/\s+/g, '+');

  return useQuery({
    queryKey: ['coupangProducts', q],
    queryFn: () => fetchCoupangProducts(q),
  });
};
